### PR TITLE
feat(sharding): add command to sync tables onto new nodes

### LINKED ISSUE #8904: Automatically propagate schema to all clickhouse nodes
## Is your feature request related to a problem?

`clickhouse-operator` does not automatically create all relevant tables on all shards when resharding. Related: https://github.com/Altinity/clickhouse-operator/issues/903

## Describe the solution you'd like

Automatically create tables on all replicas in bin/migrate

### Constraints

- Running N ON CLUSTER commands is bound to be slow in bin/migrate. A better alternative might be to use `clusterAllReplicas` function to query system.tables on each host and run CREATE TABLE IF NOT EXISTS only if table counts look off for primary tables.

## Describe alternatives you've considered

- Ask users to manually update schema.

## Additional context

Prerequisite for shipping https://github.com/PostHog/posthog/issues/8652

### PR DESCRIPTION
**Pull request by macobo**
clickhouse-operator only syncs some tables onto new nodes. This new
command ensures that when adding new shards, they are automatically
synced up on redeploying

Note that there might be timing concerns here as resharding on altinity
cloud does not redeploy automatically. In practice however what this
means is that new nodes just won't ingest any data until another deploy

Closes https://github.com/PostHog/posthog/issues/8904, related to https://github.com/PostHog/posthog/issues/8652

## How did you test this code?

Custom-built an image and deployed it in a sharded setting.

### CODE REVIEW

**Review by guidoiaquinti — APPROVED** [review]
LGTM and few non-blocking comments:

- should we add a specific test for scenarios in `sync_replicated_schema`?
- should we handle as well the case of ghost tables? (e.g. tables that were dropped on some nodes but not on all of them)

**Comment by macobo** [comment]
> should we handle as well the case of ghost tables? (e.g. tables that were dropped on some nodes but not on all of them)

Automatically dropping client data is a huge no-no in my book so not something I want to touch with a 20-foot pole. This is not an issue right now IMO.

> should we add a specific test for scenarios in sync_replicated_schema?

Not really given our current CI setup or without mocking the DB. The mocking route wouldn't really give us anything either. Given how light-weight this script is, I don't feel too uncomfortable with this as-is, but do argue if you disagree!

File: ee/management/commands/sync_replicated_schema.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            FROM clusterAllReplicas(%(cluster)s, system, tables)
+            WHERE database = %(database)s
+              AND name IN %(table_names)s
+            GROUP BY host
+        """,
+            {
+                "cluster": CLICKHOUSE_CLUSTER,
+                "database": CLICKHOUSE_DATABASE,
+                "table_names": table_names,
+                "target_count": len(table_names),
```
Comment (hazzadous) [inline ee/management/commands/sync_replicated_schema.py:55]: Is this needed? Doesn't look like it's being used, but maybe it should?

File: ee/management/commands/sync_replicated_schema.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+from django.core.management.base import BaseCommand
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.schema import CREATE_TABLE_QUERIES, build_query, get_table_name
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE, CLICKHOUSE_REPLICATION, MULTI_TENANCY
+
+logger = structlog.get_logger(__name__)
+
+
+class Command(BaseCommand):
```
Comment (hazzadous) [inline ee/management/commands/sync_replicated_schema.py:14]: Can we add a test for this command? Even just to check that it runs. I'd like to know if I break this accidentally, as it seems a dependency to running in production.

File: ee/management/commands/sync_replicated_schema.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+        if len(out_of_sync_hosts) > 0:
+            logger.info("Schema out of sync on some clickhouse nodes!", out_of_sync_hosts=out_of_sync_hosts)
+
+            if options["dry_run"]:
+                exit(1)
+
+            logger.info("Creating missing tables")
+            for query in CREATE_TABLE_QUERIES:
+                sync_execute(build_query(query))
```
Comment (hazzadous) [inline ee/management/commands/sync_replicated_schema.py:37]: I assume these are ON CLUSTER queries, and that will not cause issues if tables are already defined on a node?

**Review by hazzadous — CHANGES_REQUESTED** [review]
I think it's worth adding a test here given it's critical for production deployments that this runs as expected.

The other comments are non-blocking

File: ee/management/commands/sync_replicated_schema.py (hunk 3 above)
Comment (macobo) [reply ee/management/commands/sync_replicated_schema.py:37]: Exactly - you can also check `schema.py` for the queries.

File: ee/management/commands/sync_replicated_schema.py (hunk 1 above)
Comment (macobo) [reply ee/management/commands/sync_replicated_schema.py:55]: Oops, not needed - got lost in refactoring.

File: ee/management/commands/sync_replicated_schema.py (hunk 2 above)
Comment (macobo) [reply ee/management/commands/sync_replicated_schema.py:13]: Done so! Thanks for pushing me on this - it turns out some tables were sensitive around ordering in `CREATE_TABLE_QUERIES` array that wasn't exposed on my sharded setup (as they were automatically created).

File: ee/management/commands/test/test_sync_replicated_schema.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.schema import CREATE_TABLE_QUERIES, EVENTS_TABLE_SQL
+from ee.clickhouse.util import ClickhouseTestMixin
+from ee.management.commands.sync_replicated_schema import Command
+from posthog.test.base import BaseTest
+
+
+@pytest.mark.ee
+class Test0004ReplicatedSchema(BaseTest, ClickhouseTestMixin):
```
Comment (hazzadous) [inline ee/management/commands/test/test_sync_replicated_schema.py:12]: non-blocking

Should this be called TestSyncReplicatedSchema, or is this specific to 0004 migration or something?

**Review by hazzadous — APPROVED** [review]

**Pull request merged**